application/controllers/QuotasController.php

```diff
@@ -427,19 +402,45 @@ public function actionMassiveAction($action, $surveyid)
         if ($quotaService->checkActionPermissions($action)) {
             $sItems = Yii::app()->request->getPost('sItems', '');
             $aQuotaIds = json_decode($sItems);
-            if (isset($_POST['QuotaLanguageSetting'])) {
-                $errors = $quotaService->multipleItemsAction($aQuotaIds, $action, $_POST['QuotaLanguageSetting']);
-            } else {
-                $errors = $quotaService->multipleItemsAction($aQuotaIds, $action);
-            }
+            $errors = $quotaService->multipleItemsAction(
+                $aQuotaIds,
+                $action,
+                Yii::app()->request->getPost('QuotaLanguageSetting', [])
+            );
             if (empty($errors)) {
                 eT("OK!");
             } else {
                 eT("Error!");
             }
         } else {
+            /* 403 error ? */
             Yii::app()->user->setFlash('error', gT("Access denied."));
             $this->redirect(Yii::app()->request->urlReferrer);
         }
     }
+
+    /**
+     * Check permission on a specific quota
+     * Check survey id and quota id at same time
+     * @param integer $iSurveyId
+     * @param string $sPermission
+     * @param integer|null $iQuotaid
+     * throw Exception
+     * @return void
+     */
+    private function checkPermissions($surveyId, $sPermission = 'read', $quotaId = null)
+    {
+        if (!Permission::model()->hasSurveyPermission($surveyId, 'quotas', $sPermission)) {
+            throw new CHttpException(403, gT("You do not have permission on this survey."));
+        }
+        if ($quotaId) {
```

Comment:
Is this part of the checking permission? Or this is part of a parent (survey) - child (quota) relationship checking?
Usually these kind of checks are not done on checkPermission routines.
I would say this needs to be moved out of this method.

Reply:
>  I would say this needs to be moved out of this method.

You don't like Permission ? I can move to checkValidity …

Reply:
> You don't like Permission ? I can move to checkValidity …

I don;t like checking the quota belongs to a survey inside a checkPermission.
I believe they should be 2 different functions, and 2 different steps, as done everywhere else, right?.

Reply:
> as done everywhere else, right?.

? Where ?

Reply:
Ex:
https://github.com/LimeSurvey/LimeSurvey/blob/master/application/controllers/QuestionAdministrationController.php#L125

- The question is fetched and checked it exists.
- Then the permission is checked against the surveyid stored in the model.

This way of doing it:
- is not consistent with the rest of the code 
- I don't like the idea of checking permission with the surveyId got from a post when the quota model is available

Reply:
OK for master : 

```
$quota = $this->getQuotaWithPermission($quotaId, 'update');
$sureyid = $quota->sid;
```

But not for 5.X …

Reply:
> But not for 5.X …

sounds reasonable.

Reply:
```
    /**
     * Get a quote aftre check exist and permission on a specific quota using permission on survey
     * @param integer $quotaId
     * @param string $sPermission to check (on survey quotas)
     * throw Exception
     * @return \Quota
     */
    private function getQuotaWithPermission($quotaId, $sPermission = 'read')
    {
        $oQuota = Quota::model()->findByPk($quotaId);
        if (empty($oQuota)) {
            throw new CHttpException(404, gT("Quota not found."));
        }
        if (!Permission::model()->hasSurveyPermission($oQuota->sid, 'quotas', $sPermission)) {
            throw new CHttpException(403, gT("You do not have permission on this quota."));
        }
        return $oQuota;
    }
```

(i update phpdoc)